<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Quality-Issues" data-toc-modified-id="Quality-Issues-1">Quality Issues</a></span></li><li><span><a href="#Tidiness" data-toc-modified-id="Tidiness-2">Tidiness</a></span></li></ul></div>

In [178]:
import pandas as pd
import requests 
import re
import os
from timeit import default_timer as timer
import tweepy
import json

In [158]:
keys=pd.read_csv('Keys.csv', names=['type','key'])
consumer_key = keys['key'][0]
consumer_secret = keys['key'][1]
access_token = keys['key'][2]
access_secret = keys['key'][3]

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [182]:
data=[]
errors=[]
i=0
for id in df.tweet_id:
    start = timer()
    i+=1
    try:
        tweet = api.get_status(id, tweet_mode='extended')
        data.append(tweet._json)
    except:
        errors.append(id)
    end = timer()
    print('ID: {} - time: {}'.format(id,end-start))
    if i == 3:
        break
with open('tweet_json.txt','w') as file:
    json.dump(data,file)

ID: 892420643555336193 - time: 0.688442564001889
ID: 892177421306343426 - time: 0.6895923179981764
ID: 891815181378084864 - time: 0.7327446580020478


In [171]:
tweet = api.get_status('892420643555336193', tweet_mode='extended')
print(tweet._json)

{'created_at': 'Tue Aug 01 16:23:56 +0000 2017', 'id': 892420643555336193, 'id_str': '892420643555336193', 'full_text': "This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", 'truncated': False, 'display_text_range': [0, 85], 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'large': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}, 'extended_entities': {'media

In [2]:
df=pd.read_csv('twitter-archive-enhanced.csv')
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [96]:
78+181
2356-259

2097

In [ ]:
url="https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"

In [12]:
image_pred_file=requests.get(url)
with open(os.path.join(url.split('/')[-1]),mode='wb') as file:
        file.write(image_pred_file.content)
df_image=pd.read_csv(url.split('/')[-1], sep='\t')
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [12]:
image_pred_file=requests.get(url)
with open(os.path.join(url.split('/')[-1]),mode='wb') as file:
        file.write(image_pred_file.content)
df_image=pd.read_csv(url.split('/')[-1], sep='\t')
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [118]:
df_image.sample(10)

tweet_id                                          jpg_url  \
1046  713175907180089344  https://pbs.twimg.com/media/CeW1tERWAAAA9Q2.jpg   
1423  772117678702071809  https://pbs.twimg.com/media/Crcc7pqXEAAM5O2.jpg   
1266  749395845976588288  https://pbs.twimg.com/media/CmZjizYW8AA3FCN.jpg   
1744  822872901745569793  https://pbs.twimg.com/media/C2tugXLXgAArJO4.jpg   
964   706166467411222528  https://pbs.twimg.com/media/CczOp_OWoAAo5zR.jpg   
1482  781163403222056960  https://pbs.twimg.com/media/Ctc_-BTWEAAQpZh.jpg   
1258  748692773788876800  https://pbs.twimg.com/media/CmPkGhFXEAABO1n.jpg   
2022  881268444196462592  https://pbs.twimg.com/media/DDrk-f9WAAI-WQv.jpg   
304   671518598289059840  https://pbs.twimg.com/media/CVG2l9jUYAAwg-w.jpg   
1398  768473857036525572  https://pbs.twimg.com/media/Cqoq5PGWAAA-U8T.jpg   

      img_num                  p1   p1_conf  p1_dog                        p2  \
1046        1         timber_wolf  0.503788   False                  malamute   
1423        1  Labrador_retriever  0.217821    True                    beagle   
1266        1          Pomeranian  0.973715    True                      chow   
1744        1    Lakeland_terrier  0.196015    True        Labrador_retriever   
964         1             Samoyed  0.430418    True                    kuvasz   
1482        1   Shetland_sheepdog  0.973841    True                    collie   
1258        1                  ox  0.337871   False                      plow   
2022        1              tusker  0.473303   False           Indian_elephant   
304         1    Lakeland_terrier  0.428275    True   wire-haired_fox_terrier   
1398        1  Labrador_retriever  0.739170    True  Chesapeake_Bay_retriever   

       p2_conf  p2_dog                p3   p3_conf  p3_dog  
1046  0.430624    True    Siberian_husky  0.028454    True  
1423  0.157677    True  golden_retriever  0.127726    True  
1266  0.020758    True          keeshond  0.003784    True  
1744  0.160329    True     Irish_terrier  0.069126    True  
964   0.279600    True    Great_Pyrenees  0.117480    True  
1482  0.025188    True     Border_collie  0.000297    True  
1258  0.269287   False            oxcart  0.245653   False  
2022  0.245646   False              ibex  0.055661   False  
304   0.111472    True        toy_poodle  0.105016    True  
1398  0.246488    True            kelpie  0.006892    True

In [121]:
df.name.value_counts()

None       745
a           55
Charlie     12
Cooper      11
Oliver      11
          ... 
Sailor       1
Kayla        1
Pippin       1
Molly        1
BeBe         1
Name: name, Length: 957, dtype: int64

In [123]:
df[df.name=='a'].text

56      Here is a pupper approaching maximum borkdrive...
649     Here is a perfect example of someone who has t...
801     Guys this is getting so out of hand. We only r...
1002    This is a mighty rare blue-tailed hammer sherk...
1004    Viewer discretion is advised. This is a terrib...
1017    This is a carrot. We only rate dogs. Please on...
1049    This is a very rare Great Alaskan Bush Pupper....
1193    People please. This is a Deadly Mediterranean ...
1207    This is a taco. We only rate dogs. Please only...
1340    Here is a heartbreaking scene of an incredible...
1351    Here is a whole flock of puppers.  60/50 I'll ...
1361    This is a Butternut Cumberfloof. It's not wind...
1368    This is a Wild Tuscan Poofwiggle. Careful not ...
1382    "Pupper is a present to world. Here is a bow f...
1499    This is a rare Arctic Wubberfloof. Unamused by...
1737    Guys this really needs to stop. We've been ove...
1785    This is a dog swinging. I really enjoyed it so...
1853    This i

In [131]:
print (df.iloc[1853].text)

This is a Sizzlin Menorah spaniel from Brooklyn named Wylie. Lovable eyes. Chiller as hell. 10/10 and I'm out.. poof https://t.co/7E0AiJXPmI


In [90]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [119]:
df[df.tweet_id==676613908052996102]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1815  676613908052996102                    NaN                  NaN   

                      timestamp  \
1815  2015-12-15 04:05:01 +0000   

                                                 source  \
1815  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1815  This is the saddest/sweetest/best picture I've...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1815                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1815  https://twitter.com/dog_rates/status/676613908...                12   

      rating_denominator name doggo floofer pupper puppo  
1815                  10  the  None    None   None  None

In [120]:
df[df.tweet_id==713175907180089344]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1232  713175907180089344                    NaN                  NaN   

                      timestamp  \
1232  2016-03-25 01:29:21 +0000   

                                                 source  \
1232  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1232  Say hello to Opie and Clarkus. Clarkus fell as...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1232                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1232  https://twitter.com/dog_rates/status/713175907...                10   

      rating_denominator  name doggo floofer pupper puppo  
1232                  10  Opie  None    None   None  None

In [76]:
df.sample(4)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
477   815990720817401858                    NaN                  NaN   
1872  675146535592706048                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
769   777189768882946048                    NaN                  NaN   

                      timestamp  \
477   2017-01-02 18:38:42 +0000   
1872  2015-12-11 02:54:12 +0000   
2235  2015-11-21 20:59:20 +0000   
769   2016-09-17 16:57:35 +0000   

                                                 source  \
477   <a href="http://twitter.com/download/iphone" r...   
1872  <a href="http://twitter.com/download/iphone" r...   
2235  <a href="http://twitter.com/download/iphone" r...   
769   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
477   Meet Jack. He's one of the rare doggos that do...                  NaN   
1872  This is Coops. He's yelling at the carpet. Not...                  NaN   
2235  This is a Trans Siberian Kellogg named Alfonso...                  NaN   
769   This is Jeffrey. He's being held so he doesn't...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
477                        NaN                        NaN   
1872                       NaN                        NaN   
2235                       NaN                        NaN   
769                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
477   https://www.gofundme.com/surgeryforjacktheminp...                11   
1872  https://twitter.com/dog_rates/status/675146535...                 7   
2235  https://twitter.com/dog_rates/status/668171859...                 7   
769   https://twitter.com/dog_rates/status/777189768...                12   

      rating_denominator     name doggo floofer pupper puppo  
477                   10     Jack  None    None   None  None  
1872                  10    Coops  None    None   None  None  
2235                  10        a  None    None   None  None  
769                   10  Jeffrey  None    None   None  None

In [86]:
df.retweeted_status_id.count()

181

In [87]:
df[df.retweeted_status_id.notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
...                  ...                    ...                  ...   
1023  746521445350707200                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
32    <a href="http://twitter.com/download/iphone" r...   
36    <a href="http://twitter.com/download/iphone" r...   
68    <a href="http://twitter.com/download/iphone" r...   
73    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1023  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
1242  <a href="http://twitter.com/download/iphone" r...   
2259  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
19    RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36    RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68    RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73    RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
...                                                 ...                  ...   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
1043  RT @dog_rates: Extremely intelligent dog here....         6.671383e+17   
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. ...         7.119983e+17   
2259  RT @dogratingrating: Exceptional talent. Origi...         6.675487e+17   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
19                4.196984e+09  2017-07-19 00:47:34 +0000   
32                1.960740e+07  2017-07-15 02:44:07 +0000   
36                4.196984e+09  2017-02-12 01:04:29 +0000   
68                4.196984e+09  2017-06-23 01:10:23 +0000   
73                4.196984e+09  2017-06-23 16:00:04 +0000   
...                        ...                        ...   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
1043              4.196984e+09  2015-11-19 00:32:12 +0000   
1242              7.832140e+05  2016-03-21 19:29:52 +0000   
2259              4.296832e+09  2015-11-20 03:43:06 +0000   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   

                                          expanded_urls  rating_numerator  \
19    https://twitter.com/dog_rates/status/887473957...                13   
32    https://twitter.com/dog_rates/status/886053434...                12   
36    https://twitter.com/dog_rates/status/830583320...                13   
68    https://twitt

In [44]:
df.timestamp.min()

'2015-11-15 22:32:08 +0000'

In [45]:
df.timestamp.max()

'2017-08-01 16:23:56 +0000'

In [40]:
df.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [113]:
df.text.str.contains('doggo',case=False).sum()

107

In [41]:
df.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [114]:
df.text.str.contains('floofer',case=False).sum()

10

In [42]:
df.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [115]:
df.text.str.contains('pupper',case=False).sum()

283

In [43]:
df.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [116]:
df.text.str.contains('puppo',case=False).sum()

38

In [35]:
for index, tweet in df[df.rating_denominator>10].iterrows():
    print ("{} - Text: {} - {}/{}".format(index,tweet.text,tweet.rating_numerator,tweet.rating_denominator))

342 - Text: @docmisterio account started on 11/15/15 - 11/15
433 - Text: The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd - 84/70
784 - Text: RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/… - 9/11
902 - Text: Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE - 165/150
1068 - Text: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ - 9/11
1120 - Text: Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv - 204/170
1165 - Text: Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a - 4/20
1202 - Text: This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq - 50/50
1228 - Text: Happy Saturday 

In [109]:
df[['text','retweeted_status_id']].iloc[1068]

text                   After so many requests, this is Bretagne. She ...
retweeted_status_id                                                  NaN
Name: 1068, dtype: object

In [110]:
df[['text','retweeted_status_id']].iloc[784]

text                   RT @dog_rates: After so many requests, this is...
retweeted_status_id                                          7.40373e+17
Name: 784, dtype: object

In [111]:
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [47]:
df[df.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [71]:
len(str(df.tweet_id[8]))

18

In [75]:
for index, tweet in df.iterrows():
    try:
        length=len(str(tweet.tweet_id))
        if length<18:
            print ("{} - lenght: {}".format(index,length))
    except TypeError:
        print ("{} - type error".format(index))
    #else:
     #   print ("{} - error".format(index))

In [37]:
df.duplicated().sum()

0

In [38]:
df_image.duplicated().sum()

0

## Quality Issues

- Dog stages data type should be categorical
- Timestamp should be Date/Time data type
- Fix score for the following tweets:
    - 784: Score is not accurate since it was confused with 9/11. Correct score is 14/10
    - 1165 - 4/20 is a date, fix to 13/10
    - 1202 - should be 11/10 - 50/50 is a split
    - 1662 - 10/10 instead of 7/11
- Remove no dogs related tweets: Indexes: 342, 433,1598,1663
- Adjust missing dog stages:
    - 10 doggo tweets seems to be missing
    - 26 floofer
    - 8 puppo
    - Identify wrong categories
- Adjust name
    - 1232: two names Opie and Clarkus
    - 1853: Wylie

## Tidiness

- Dog stages should be only one column
- Remove retweets: Retweets are classified as those records with `retweeted_status_id`
- Remove unneeded columns and add retweets and favorite counts as well as best match


In reply to must be integer
expanded url is missing from 59 post
doggo floofer pupper puppo could be all in one column 
new column type: categorical
remove retweets
remove no dogs tweets
time to date_time
Updated expanded URL for 59 missing cases
check expanded URL 734. Repeat


check: 
342 need to be removed
433
784: Score is not accurate since it was confused with 9/11. Correct score is 14/10
784 and 1068 seems to be duplicated
1165 - 4/20 is a date, fix to 13/10
1202 - should be 11/10 - 50/50 is a split
check 1598
1662 - 10/10 instead of 7/11
check 1663



Visualization

Retweets - Favorite vs Rating (num/den)
Type of dog vs rating
accuracy of algorithms for dog type: match best type with text (lowercase) to see if the algorithm is good.
